README
Bio::Chado::Schema - A DBIx::Class ORM layer for Chado

This is a standard object-relational mapping layer for use with the GMOD Chado
database schema.  This layer is implemented with DBIx::Class, generated with
the help of the very fine DBIx::Class::Schema::Loader module.

Chado is an open-source modular database schema for biological data.  It is
divided into several notional "modules", which are reflected in the namespace
organization of this package.  Note that modules in the Chado context refers to
sets of tables, they are not modules in the Perl sense.

Installation:

If you are using a distribution tarball of Bio::Chado::Schema (such as from CPAN) :

    perl Build.PL

    ./Build installdeps

    ./Build test

    ./Build install

If you are developer using the git repo, then Dist::Zilla needs to be used:

    dzil build

    dzil test

    dzil install


Changes
{{$NEXT}}

0.10010   2011-11-08 16:39:15 America/New_York

 * fix (in cleanup) errors related to schema object DESTROY function
   on some versions of DBIC
 * added create_phenotypeprop function (naama)
 * fix filenames of Chado-module-level POD files.  For instance,
   Bio::Chado::Schema::Result::Sequence POD is now rightly in
   lib/Bio/Chado/Schema/Result/Sequence.pod

0.10000   2011-11-07 13:22:41 America/Los_Angeles

 * add support for phenotypeprop table from Chado svn r25232

0.09040   2011-11-04 09:31:11 America/Los_Angeles

 * fix a memory-leak bug present in 0.9010 through 0.9030

0.09030   2011-11-03 14:58:58 America/Los_Angeles

 * fix some issues with repeated sections in the POD

0.09020   2011-11-02 17:30:03 America/Los_Angeles

 * implemented alphabet() function for Feature rows using as a silly
   heuristic that works in many cases.  alphabet() is needed by things
   like revcom() in newer versions of BioPerl.

0.09010   2011-10-24 11:25:25 America/Los_Angeles

 * added get_cvterm and get_cvterm_or_die convenience methods directly
   to the schema object, since cvterms are so pervasive in chado
 * added to_range to Featureloc class that makes a Bio::Range object
   for that location.

0.09000   2011-09-22 09:27:36 America/Los_Angeles

 * update for Chado 1.2, including:
    + add classes for new Chado 1.2 tables chadoprop (Cv::Chadoprop)
      genotypeprop (Genetic::Genotypeprop)
    + adds type to nd_protocol (NaturalDiversity::NdProtocol)
    + adds name column to phenotype table (Phenotype::Phenotype)
    + corrects misspelled name of stock_relationship_id column
      accessor for stock_relationship_cvterm
      (Stock::StockRelationshipCvterm), adds stock_relationship DBIC
      relation

0.08200   2011-09-19 09:35:03 America/Los_Angeles

 * fix inconsistencies in behavior for features with null residues
   and/or seqlens

0.08100   2011-03-21 16:01:00 America/Los_Angeles

 * add support for literal sql in property values set using
   create_*props helper functions
 * add 'features' many-to-many relation to the Synonym result class.
 * add 'synonyms' many-to-many relation to the Feature result class.
 * add new class for stock_cvtermprop table, and new is_not and rank
   columns for stock_cvterm

0.08002   2011-02-22 11:17:13 America/Los_Angeles

 * Added support for new cvprop table.
 * Added plugin_add_relationship class method to Bio::Chado::Schema
   main class, which makes it possible to do BCS plugins: sets of
   classes in the Bio::Chado::Schema namespace that are loaded as part
   of BCS.  plugin_add_relationship lets plugins inject relationships
   into the core BCS classes.

0.07300   2010-11-28 17:19:08 PST8PDT

 * Add convenient featureloc length() method.
 * Add better support for large sequences: a subseq() method to
   Feature class with support for checking a 'large_residues'
   featureprop for sequence, and doing substring calls on that.

0.07100   2010-11-16 12:28:17 PST8PDT

 * Fix bug with Cvterm add_secondary_dbxref, only caught under
   DBIx::Class >= 0.08124
 * add convenience methods for creating stock_dbxrefprops,
   nd_experimentprops, projectprops, and geolocationprops
 * add methods for working with parent and child terms of cvterms
 * Convert t/lib/BCSTest.pm to Bio::Chado::Schema::Test so other
   CPAN modules can have test suites that use an
   automagically-created SQLite db

0.06400   2010-10-20 20:54:37 PST8PDT

 * added naama's stock_dbxrefprop from upstream chado

0.06302   2010-10-20 09:58:19 PST8PDT

 * corrected homepage url in CPAN upload

0.06301   2010-10-19 17:49:41 PST8PDT

 * corrected missing DBIx::Class::Tree::NestedSet dep

0.06300   2010-10-18 17:38:29 PST8PDT

 * converted build system to Dist::Zilla
 * applied nmenda's type_id corrections from chado trunk
 * fix permission denied test failure caused by shipping a
   t/var/BCS.db file by accident

0.06200

 * removed erroneous duplicate General/Project.pm module
 * Updated natdiv module from Chado upstream (fixed upstream by nmenda).
 * Added create_stockprops method to stock module (nmenda).

0.06001

 * Corrected missing stuff in the manifest.

0.06000

 * Added DBIx::Class::Tree::NestedSet support to for Phylonodes.
 * Added Natural Diversity and Project module support.

0.05801

 * Fix bugs and add tests for child_features and parent_features on
   Sequence::Feature
 * Added parent_relationships and child_relationships rels to
   Sequence::Feature as more intuitive aliases for
   feature_relationship_*

0.05800
 
 * New aliases on Sequence::Feature
    child_features  -> featureloc_features
    parent_features -> featureloc_srcfeatures
 * Added a function for creating cvtermsynonyms ($cvterm->add_synonym)
 * Added a function for creating new cvtermprops ($cvterm->create_cvtermprops) 
 * Test suite can now deploy a Chado schema to a SQLite database
   if BCS_TEST_DSN is not defined. This means that smokers can
   now run these tests without setting up any databases.
        -- leto

0.05703

 * relaxed dependency on Carp 1.08

0.05702

 * that upload was a bit TOO clean, was missing the Build.PL!

0.05701

  * corrected a dirty cpan upload

0.05700

  * re-dumped with latest schema loader devel version
  * stripped _id suffixes for a number of relationships that still had
    them.  Hopefully the schema loader patch used for this will make
    it into schema loader releases.  List of changed relationship names:

     lib/Bio/Chado/Schema/CellLine/CellLine.pm
      cell_line_relationship_subject_ids -> cell_line_relationship_subjects
      cell_line_relationship_object_ids  -> cell_line_relationship_objects

     lib/Bio/Chado/Schema/Contact/Contact.pm
      contact_relationship_object_ids -> contact_relationship_objects
      contact_relationship_subject_ids -> contact_relationship_subjects

     lib/Bio/Chado/Schema/Cv/Cvterm.pm
      arraydesign_platformtype_ids -> arraydesign_platformtypes
      arraydesign_substratetype_ids -> arraydesign_substratetypes

lib/Bio/Chado/Schema.pm
package Bio::Chado::Schema;

# Created by DBIx::Class::Schema::Loader
# DO NOT MODIFY THE FIRST PART OF THIS FILE

use strict;
use warnings;

use base 'DBIx::Class::Schema';

__PACKAGE__->load_namespaces;


# Created by DBIx::Class::Schema::Loader v0.07010 @ 2011-03-16 23:09:58
# DO NOT MODIFY THIS OR ANYTHING ABOVE! md5sum:JJ2AbsZoAN4cnM4vrYOxKA

use Carp::Clan qr/^Bio::Chado::Schema/;
use Bio::Chado::Schema::Util;

=head1 NAME

Bio::Chado::Schema - A standard DBIx::Class layer for the Chado database schema.

=head1 SYNOPSIS

  use Bio::Chado::Schema;

  my $chado = Bio::Chado::Schema->connect( $dsn, $user, $password );

  print "number of rows in feature table: ",
        $chado->resultset('Sequence::Feature')->count,
        "\n";


=head1 DESCRIPTION

This is a standard object-relational mapping layer for use with the
GMOD Chado database schema.  This layer is implemented with
L<DBIx::Class>, generated with the help of the very fine
L<DBIx::Class::Schema::Loader> module.

Chado is an open-source modular database schema for biological data.
It is divided into several notional "modules", which are reflected in
the namespace organization of this package.  Note that modules in the
Chado context refers to sets of tables, they are not modules in the
Perl sense.

=head1 GETTING STARTED

To learn how to use this DBIx::Class ORM layer, a good starting
point is the L<DBIx::Class::Manual>.

=head1 CHADO MODULES COVERED BY THIS PACKAGE

L<Bio::Chado::Schema::CellLine>

L<Bio::Chado::Schema::Companalysis>

L<Bio::Chado::Schema::Composite>

L<Bio::Chado::Schema::Contact>

L<Bio::Chado::Schema::Cv>

L<Bio::Chado::Schema::Expression>

L<Bio::Chado::Schema::General>

L<Bio::Chado::Schema::Genetic>

L<Bio::Chado::Schema::Library>

L<Bio::Chado::Schema::Mage>

L<Bio::Chado::Schema::Map>

L<Bio::Chado::Schema::NaturalDiversity>

L<Bio::Chado::Schema::Organism>

L<Bio::Chado::Schema::Phenotype>

L<Bio::Chado::Schema::Phylogeny>

L<Bio::Chado::Schema::Project>

L<Bio::Chado::Schema::Pub>

L<Bio::Chado::Schema::Sequence>

L<Bio::Chado::Schema::Stock>


=head1 CHADO VERSIONS

Basically, BCS has always followed the SVN HEAD of Chado, since it's
on a much faster release cycle than Chado itself.

Most users will not experience incompatibilities using the most recent
versions of Bio::Chado::Schema with older Chado installations.
However, if you encounter problems using Bio::Chado::Schema with an
older Chado schema, you may want to try downgrading to an earlier
version of Bio::Chado::Schema.  The following rough equivalencies hold
between BCS and Chado versions:

  Chado 1.2   --  BCS 0.09000
  Chado 1.11  --  BCS 0.05801
  Chado 1.1   --  BCS 0.05801
  older       --  BCS 0.03100

=head1 SCHEMA OBJECT METHODS

=head2 get_cvterm( "$cv_name:$cvterm_name" ) OR get_cvterm( $cv_name, $cvterm_name )

Convenience method to for finding single cvterms based on the text
name of the CV and the term.  The cvterm objects found with this
method are cached in the schema object itself.  Thus, you only use
this function in the (relatively common) scenario in which you just
need convenient access to a handful of different cvterms.

=cut

sub get_cvterm {
    my ( $self, $cv_name, $term_name ) = @_;

    croak "must provide at least one argument!" unless @_ > 1;

    unless( $term_name ) {
        ($cv_name, $term_name) = split /:/, $cv_name, 2;
    }

    return $self->{_bio_chado_schema_cvterm_cache}{$cv_name}{$term_name} ||=
        $self->resultset('Cv::Cv')
             ->search({ 'me.name' => $cv_name })
             ->search_related('cvterms', { 'cvterms.name' => $term_name })
             ->single;
}

=head2 get_cvterm_or_die

Same as get_cvterm above, but dies with a "not found" message if the
cvterm is not found.  This is convenient when you don't want to be
bothered with checking the return value of C<get_cvterm>, which for me
is most of the time.

=cut

sub get_cvterm_or_die {
    shift->get_cvterm( @_ ) or croak "cvterm @_ not found";
}

=head1 CLASS METHODS

=head2 plugin_add_relationship( 'ChadoModule::SourceName', 'reltype', @args )

Sometimes application-specific plugins need to add relationships to
the core BCS classes.  It can't just be done normally from inside the
classes of the plugins, you need to use this method.

Example: Bio::Chado::Schema::Result::MyApp::SpecialThing belongs_to
the core BCS Organism::Organism, and you would like to be able to call
C<$organism-E<gt>myapp_specialthings> on organisms to get their
associated SpecialThings.

    package Bio::Chado::Schema::MyApp::Result::SpecialThing;

    # ( do table and column definitions and so forth here )

    Bio::Chado::Schema->plugin_add_relationship(
        'Organism::Organism', 'has_many', (
            "myapp_specialthings",
            "Bio::Chado::Schema::MyApp::Result::Foo",
            { "foreign.organism_id" => "self.organism_id" },
            { cascade_copy => 0, cascade_delete => 0 },
        );
    );

=cut

{
    my @after_load;
    $_->() for @after_load; #< note that this executes after load_classes above

    sub plugin_add_relationship {
        my ( $class, $target_moniker, $reltype, @args ) = @_;

        push @after_load, sub {
            no strict 'refs';
            my $target_class = $class->class( $target_moniker );
            $target_class->$reltype( @args );
            __PACKAGE__->register_class( $target_moniker => $target_class );
        };
    }
}

sub DESTROY {
    my $self = shift;

    # need to delete our cvterm cache to avoid memory leaks
    delete $self->{_bio_chado_schema_cvterm_cache};
    $self->SUPER::DESTROY( @_ ) if $self->can( 'SUPER::DESTROY' );
}

=head1 AUTHOR

Robert Buels, <rmb32@cornell.edu>

=head1 CONTRIBUTORS

Aureliano Bombarely, <ab782@cornell.edu>

Naama Menda, <nm249@cornell.edu>

Jonathan "Duke" Leto, <jonathan@leto.net>

=head1 COPYRIGHT & LICENSE

Copyright 2009 Boyce Thompson Institute for Plant Research

This program is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut

1;
